csrc/optimization/remove_empty.cpp

```diff
@@ -0,0 +1,493 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <optimization/remove_empty.h>
+
+#include <ir/utils.h>
+#include <ops/alias.h>
+#include <ops/arith.h>
+
+#include <algorithm>
+#include <limits>
+#include <unordered_set>
+#include <vector>
+
+namespace nvfuser::optimization {
+
+namespace {
+
+//! This is a generic traversal class that is used to modify a Fusion graph by
+//! replacing TensorViews so that their definitions can be altered to simplify
+//! computation or remove dead code. This differs from OptOutMutator, which is
+//! built for mutating TensorViews in a graph, and does not easily handle
+//! modifying TensorView definitions and Expr Fusion inputs during traversal.
+//!
+//! We use unordered_set called live_statements_, which is initialized as the
+//! Exprs in traversal_exprs_ as well as their inputs and their outputs with
+//! live uses. Marking a Statement as dead removes it from live_statements_,
+//! and replacing a Val inserts the Val and its definition, recursively. Since
+//! we traverse backwards, and we handle all active Expr outputs, this ensures
+//! that it is safe to removing an Expr will not result in erasing definitions
+//! of active Expr outputs.
+//!
+//! Derived classes should override handle() and make use of replaceTV(),
+//! markDeadAndMaybeRemove(), and allUsesDead(). Note that if replacements are
+//! made using replaceTV(old_tv, new_tv), then neither new_tv or any new
+//! Statements produced in creating it will be traversed by this class.
+class DeadCodeRemover : BackwardVisitor {
+ public:
+  DeadCodeRemover(Fusion* fusion) : BackwardVisitor(false), fusion_(fusion) {}
+
+  void run() {
+    // First we build a set of all live Statements so that we can detect dead
+    // branches.
+    auto exprs = StmtSort::getExprs(fusion_, fusion_->outputs());
+    // Mark every Expr, as well as its inputs and outputs as live initially
+    for (auto expr : exprs) {
+      markLive(expr);
+      for (auto inp : expr->inputs()) {
+        markLive(inp);
+      }
+      for (auto outp : expr->outputs()) {
+        markLive(outp);
+      }
+    }
+
+    // We do not traverse all outputs of all Exprs, since this requires that all
+    // paths lead to fusion_->outputs(). Instead, here we mark any Vals dead
+    // that do not have any live uses. After this, it is safe to check whether
+    // all outputs of an Expr are marked dead to determine if it should be dead,
+    // which helps us determine when it is safe to remove a multi-output
+    // definition of a dead TV.
+    std::vector<Statement*> dead_expr_outputs;
+    for (auto stmt : live_statements_) {
```

Comment:
I'm wondering if this is too conservative and would miss dead branches.

For a chained operations, `tv1 = op1(tv0); tv2 = op2(tv1)`, if we realize that tv2 has `allUsesDead` and push it to dead_expr_outputs, we can mark op2 dead as well, assuming it has no other uses.
But since we defer `markDead` after traversal here, we would miss the whole branch other than the leaf node.

Also, the order of iterating `live_statements_` is important here. but we have `std::unordered_set<Statement*> live_statements_;`

Reply:
I forgot that we are inside a BackwardVisitor, so as long as we marked the leaf node as dead, the traversal at the end should have propagated this for us.

Reply:
Sorry, the comment probably should be clearer. The order of this loop should not matter since only `Val`s and not `Expr`s are marked dead here, so it won't change `allDeadUses`. I added this loop because of `WelfordOp`, which has three outputs: `avg`, `var`, `n`.  In the empty Welford test, I make use of `avg` and `var`, but not `n`. However, in the loop before this one, we go over all `Expr`s in topological order and add mark the ops and all of their inputs and outputs as alive, so `n` gets marked live. The loop here just goes and removes those with all dead uses.

But now that I type this out I realize it's really dumb, and all I need to do is use `StmtSort::getStmts` in the first loop then I can avoid all this awkwardness. I just pushed the change. Thanks for pointing this out!